dkg/pedersen/handler.go

```diff
@@ -933,6 +941,143 @@ func (h *Handler) receiveDealsResharing(ctx context.Context, isCommonNode bool,
 	return nil
 }
 
+func (h *Handler) handleVerifiableDecrypt(out mino.Sender,
+	msg types.VerifiableDecryptRequest, from mino.Address) error {
+
+	type job struct {
+		index int // index to put the response
+		ct    types.Ciphertext
+	}
+
+	ciphertexts := msg.GetCiphertexts()
+	batchsize := len(ciphertexts)
+
+	wgBatchReply := sync.WaitGroup{}
+
+	shareAndProofs := make([]types.ShareAndProof, batchsize)
+	jobChan := make(chan job)
+
+	// Fill the chan with jobs
+	go func() {
+		for i, ct := range ciphertexts {
+			jobChan <- job{
+				index: i,
+				ct:    ct,
+			}
+
+		}
+		close(jobChan)
+	}()
+
+	n := workerNum
+	if batchsize < n {
+		n = batchsize
+	}
+
+	// Spin up workers to process jobs from the chan
+	for i := 0; i < n; i++ {
+		wgBatchReply.Add(1)
+		go func() {
+			defer wgBatchReply.Done()
+
+			for j := range jobChan {
+				sp, err := h.verifiableDecryption(j.ct)
+				if err != nil {
+					h.log.Err(err).Msg("verifiable decryption failed")
+				}
+
+				shareAndProofs[j.index] = *sp
+			}
+
+		}()
+	}
+
+	wgBatchReply.Wait()
+	verifiableDecryptReply := types.NewVerifiableDecryptReply(shareAndProofs)
+
+	errs := out.Send(verifiableDecryptReply, from)
+	err := <-errs
+	if err != nil {
+		return xerrors.Errorf("failed to send verifiable decrypt: %v", err)
+	}
+
+	return nil
+}
+
+// checkEncryptionProof verifies the encryption proofs.
+//
+// See https://arxiv.org/pdf/2205.08529.pdf / section 5.4 Protocol / step 3
+func (h *Handler) checkEncryptionProof(cp types.Ciphertext) error {
```

Comment:
I propose to put this as a library in `kyber` - then it might also be useful for other people!